File: 0143_Reorder_List/solution.java
package cn.huang.leetcode;

import cn.huang.leetcode.common.ListNode;

/*
143. Reorder List

Given a singly linked list L: L0→L1→…→Ln-1→Ln,
reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…

You may not modify the values in the list's nodes, only nodes itself may be changed.

Example 1:

Given 1->2->3->4, reorder it to 1->4->2->3.

Example 2:

Given 1->2->3->4->5, reorder it to 1->5->2->4->3.


 */
public class LC_0143_ReorderList {
    /*
    / 1. divide the linked list into two parts
    // 2. reverse the right part of the linked list
    // 3. combine two linked list one by one together

     */
    public void reorderList(ListNode head) {
        if (head == null || head.next == null)
        {
            return;
        }

        // step 1
        ListNode slow = head;
        ListNode fast = head;
        while (fast != null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }

        // step 2, remember to set slow.next to null!
        ListNode rightPart = reverseList(slow.next);
        slow.next = null;

        // step 3, combine
        ListNode leftPart = head;
        while (leftPart != null && rightPart != null)
        {
            ListNode next = leftPart.next;
            leftPart.next = rightPart;
            rightPart = rightPart.next;
            leftPart.next.next = next;
            leftPart = next;
        }
    }
    private ListNode reverseList(ListNode head)
    {
        if (head == null || head.next == null)
        {
            return head;
        }

        ListNode pre = head;
        ListNode curr = head.next;

        while (curr != null)
        {
            ListNode temp = curr.next;
            curr.next = pre;
            pre = curr;
            curr = temp;
        }

        // set head node's next
        head.next = null;

        return pre;
    }
}

File: 0143_Reorder_List/solution.cpp
/*
143. Reorder List

Level: Medium

https://leetcode.com/problems/reorder-list
*/
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <numeric>
#include <algorithm>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{

private:
  ListNode *reverseList(ListNode *head)
  {
    if (!head)
    {
      return head;
    }
    ListNode *prev = nullptr;
    ListNode *curr = head;
    while (curr)
    {
      ListNode *next = curr->next;
      curr->next = prev;
      prev = curr;
      curr = next;
    }
    return prev;
  }

  ListNode *mergeList(ListNode *l1, ListNode *l2)
  {
    ListNode dummyNode(0);
    ListNode *dummy = &dummyNode;

    bool use1 = true;
    while (l1 && l2)
    {
      if (use1)
      {
        dummy->next = l1;
        l1 = l1->next;
      }
      else
      {
        dummy->next = l2;
        l2 = l2->next;
      }
      dummy = dummy->next;
      use1 = !use1;
    }
    if (l1)
    {
      dummy->next = l1;
    }
    if (l2)
    {
      dummy->next = l2;
    }
    return dummyNode.next;
  }

public:
  void reorderList(ListNode *head)
  {
    if (!head || !head->next)
    {
      return;
    }
    ListNode *slow = head;
    ListNode *fast = head->next;

    while (fast && fast->next)
    {
      slow = slow->next;
      fast = fast->next->next;
    }
    ListNode *head2 = slow->next;
    slow->next = nullptr;

    head2 = reverseList(head2);

    mergeList(head, head2);
  }
};